Migrate stylelint to latest version #2292

LamJiuFong opened the pull request:
**What is the purpose of this pull request?**
Fixes #2292 
- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [ ] Code maintenance
- [x] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: #xxx

  If this pull request completely addresses an issue, use one of the closing keywords: "Fixes #xxx" or "Resolves #xxx"
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Migrate stylelint to latest version ^16.0.0
Add devdependency "stylelint-config-recommended-vue"
Add plugin @stylistic/stylelint-plugin 
Add dependency postcss@^8.4.35
Fix syntax to obey new rules

**Anything you'd like to highlight/discuss:**
Added a comment below

**Testing instructions:**

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Migrate stylelint to v16.0.0

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [ ] Linked all related issues
- [ ] No unrelated changes <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

---

**Reviewer checklist:**

Indicate the [SEMVER](https://semver.org/) impact of the PR:
- [ ] Major (when you make incompatible API changes)
- [ ] Minor (when you add functionality in a backward compatible manner)
- [x] Patch (when you make backward compatible bug fixes)

At the end of the review, please label the PR with the appropriate label: `r.Major`, `r.Minor`, `r.Patch`.

Breaking change release note preparation (if applicable):
- To be included in the release note for any feature that is made obsolete/breaking

> Give a brief explanation note about:
>   - what was the old feature that was made obsolete
>   - any replacement feature (if any), and
>   - how the author should modify his website to migrate from the old feature to the replacement feature (if possible).

LamJiuFong commented:
Hi, I am currently working on this issue locally and I would like to ask for some advices:

This is what I've done:
1.  Update the dependencies to the latest versions and add the dependency `stylelint-config-recommended-vue` since stylelint doesn't support linting vue files and it is also a recommended config by stylelint.<img width="427" alt="Screenshot 2024-02-15 at 2 35 19 PM" src="https://github.com/MarkBind/markbind/assets/122192553/2472e84e-50bb-4f8f-9abc-2ee68f9d1ea6">

2. Add `stylelint-config-recommended-vue` to the extension list and removed (commented out) the `indentation` rule since it is not supported by stylelint. <img width="688" alt="Screenshot 2024-02-15 at 2 39 30 PM" src="https://github.com/MarkBind/markbind/assets/122192553/34a7d13f-6d07-4fe0-a18f-899db9411287">

3. Run `npm install` followed by `npm run csslintfix` and got the errors below:
<img width="831" alt="Screenshot 2024-02-15 at 2 42 14 PM" src="https://github.com/MarkBind/markbind/assets/122192553/5f17c693-d7cd-4899-9258-d5ad094107ff">

4. Manually fix the errors above, no errors left.
5. Run `npm run updatetest` and `npm run test`. Successfully ran target test.

My questions are:
1. Is it the right way to solve this issue by following the steps above? Did I miss out anything?
2. Do I need to change anything in the `.github/workflows/ci.yml` file? If not, is it because Github Actions will run the `ci.yml` in my code which uses the updated dependencies.
3. Some new rules are introduced and running `npm run csslintfix` changes the original code.
eg. `border-color: rgba(150, 150, 150, 0.2)` is changed to `border-color: rgb(150 150 150 / 20%)`. Is it okay to keep the new changes? I think it should be fine as long as the syntax is standardised everywhere.
4. Not sure why some css files are included in `.stylelintignore`, could someone explain the purpose of it? Thanks!
<img width="543" alt="Screenshot 2024-02-15 at 8 54 18 PM" src="https://github.com/MarkBind/markbind/assets/122192553/ea5c1a30-1240-4ae7-b049-5be243536440">

EltonGohJH commented:
> 1. Is it the right way to solve this issue by following the steps above? Did I miss out anything?<br>

Looks correct on my end. For the indentation rule can you check on whether we can get a replacement? 
> 2. Do I need to change anything in the `.github/workflows/ci.yml` file? If not, is it because Github Actions will run the `ci.yml` in my code which uses the updated dependencies.

I dun think so? We still run `npm run test` command  in CI? (There is no change right?) If any change, you should change our test command.

> 3. Some new rules are introduced and running `npm run csslintfix` changes the original code.
>    eg. `border-color: rgba(150, 150, 150, 0.2)` is changed to `border-color: rgb(150 150 150 / 20%)`. Is it okay o keep the new changes? I think it should be fine as long as the syntax is standardised everywhere.

I agree that if you can fix and ensure that this new style is standardised in current code base then it is good to go.


> 4. Not sure why some css files are included in `.stylelintignore`, could someone explain the purpose of it? Thanks!
1. min.css is ignored because it is minified and will not follow the standards
2. _site is the css in templates and should not be linted.
3. coverage and test should not be linted
4. Vue components folder also has coverage folder. Maybe you can specify just the coverage folder? I dun think our vue components folder has css files. (Double check)

LamJiuFong commented:
> Looks correct on my end. For the indentation rule can you check on whether we can get a replacement?

I have added a plugin `@stylistic/stylelint-plugin` to solve this issue as they support the deprecated `indentation` rule.

> I dun think so? We still run `npm run test` command in CI? (There is no change right?) If any change, you should change our test command.

Yup, there is no change

> I agree that if you can fix and ensure that this new style is standardised in current code base then it is good to go.

Ok

> 4. Vue components folder also has coverage folder. Maybe you can specify just the coverage folder? I dun think our vue components folder has css files. (Double check)

Yes, I checked and there is no css files in the `vue-components` folder. 


Something to add on:

I ran to this error when running `npm run csslint`:

SyntaxError: Named export 'Input' not found. The requested module 'postcss' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'postcss';
const { rule: _rule, Input } = pkg;

This is because `@stylistic/stylelint-plugin` uses named export for `Input` and `postcss` only supports this starting from v8,
<img width="350" alt="Screenshot 2024-02-25 at 7 42 07 PM" src="https://github.com/MarkBind/markbind/assets/122192553/54ef35e6-e908-4b06-97ef-612f136eaa9a"> 

So, I added a dependency to specify the version of `postcss` to `^8.4.35` to get over this issue.

kaixin-hc reviewed (approved):
LGTM - I don't love how many changes there are, but I suppose it is somewhat expected with a linting tool. Will wait before merging in case I've missed something.

LamJiuFong commented on packages/core-web/src/styles/markbind.css:
```diff
@@ -27,8 +27,7 @@ pre.hljs > code {
     background: none;
 }
 
-pre > code.hljs {
```
Yup, it violates the `property-no-vendor-prefix` rule.
<img width="723" alt="Screenshot 2024-02-25 at 8 15 11 PM" src="https://github.com/MarkBind/markbind/assets/122192553/ccbbae1a-4d7b-4f56-bc80-aed66f55e779">

We can solve this issue by ignoring the property.
<img width="707" alt="Screenshot 2024-02-25 at 8 28 32 PM" src="https://github.com/MarkBind/markbind/assets/122192553/43a7e426-e173-4e54-af67-5c922ca917dc">

This will keep the original line 
`-webkit-background-clip: padding-box; /* for Safari */`

I think we can do this, what do you think?

kaixin-hc replied:
The understanding I have is that when this code was written, the browsers required prefixing for these due to not supporting the properties unprefixed. Stylelint relies on autoprefixer, an open source project that automatically adds prefixes when needed.

The implication here is that the prefixes are no longer needed due to them being supported by the browser at this point, though they were not supported in the past.

Hence, assuming there is no change in behaviour, I think we should keep the rule and change the CSS.

kaixin-hc replied:
Oh! However do remove the comments "for safari"  / references to other browsers, as these are no longer relevant.

LamJiuFong replied:
Oh I think you are right, for now since browsers do not require vendor prefixes, I think the line `background-clip: padding-box` would be enough to accommodate all browsers.

I have removed the unnecessary comments.

kaixin-hc commented:
Would you also add "Fixes #2292" to the PR description? This allows us to auto-close the issue and links the issue in the body so it easier for the reviewers!

LamJiuFong commented:
Added!

yucheng11122017 commented on packages/vue-components/src/Tooltip.vue:
```diff
@@ -71,9 +71,9 @@ export default {
 
 <style>
-    .v-popper--theme-tooltip .v-popper__inner {
+    .v-popper-theme-tooltip .v-popper-inner {
```
This seems to be having unexpected effects - the styling doesn't apply to the tooltip anymore. 
You can try modifying the font size to something huge and then removing the one of the dashes, you realise the font change doesn't apply. 
Can we keep the old version? Will it automatically change every time the linter is run?

LamJiuFong replied:
<img width="835" alt="Screenshot 2024-02-28 at 9 19 24 PM" src="https://github.com/MarkBind/markbind/assets/122192553/67ac00e6-abff-44af-b6fa-24ebbb8c4d6f">

The namings here violated the rule `selector-class-pattern`. 
To revert it back to the old version and prevent stylelint from throwing an error, I have added a line comment on top of it to disable stylelint for this part. (this is how stylelint is disabled for a particular part of code)

yucheng11122017 reviewed (commented):
Hi @LamJiuFong thank you very much for the good work! 
One minor nit and one issue with the tooltip styling

LamJiuFong commented on package.json:
```diff
@@ -1,7 +1,9 @@
-  "workspaces": [ "packages/*" ],
+  "workspaces": [
+    "packages/*"
+  ],
```
Apologies, I think my IDE has formatted this on its own and I wasn't aware of this.

yucheng11122017 reviewed (approved):
LGTM! Thanks for the work @LamJiuFong :)
